controllers/certificatesigningrequests_controller.go

```diff
@@ -46,18 +62,109 @@ type CertificateSigningRequestsReconciler struct {
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
 
-	return ctrl.Result{}, nil
+	return ctrl.Result{}, r.reconcileCSR(certificateSigningRequest, configMap)
+}
+
+// NewCertificateSigningRequestsReconciler returns a pointer to certificateSigningRequestsReconciler
+func NewCertificateSigningRequestsReconciler(mgr manager.Manager, watchNamespace string) (*certificateSigningRequestsReconciler, error) {
+	k8sClientSet, err := kubernetes.NewForConfig(mgr.GetConfig())
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating kubernetes clientset")
+	}
+
+	return &certificateSigningRequestsReconciler{
+		client:         mgr.GetClient(),
+		log:            ctrl.Log.WithName("controllers").WithName("CertificateSigningRequests"),
+		k8sclientset:   k8sClientSet,
+		watchNamespace: watchNamespace,
+	}, nil
 }
 
 // SetupWithManager sets up the controller with the Manager.
-func (r *CertificateSigningRequestsReconciler) SetupWithManager(mgr ctrl.Manager) error {
+func (r *certificateSigningRequestsReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	certificateSigningRequestsPredicate := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return pendingCertFilter(e.Object)
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return pendingCertFilter(e.ObjectNew)
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return pendingCertFilter(e.Object)
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool { return false },
+	}
 	return ctrl.NewControllerManagedBy(mgr).
-		// Uncomment the following line adding a pointer to an instance of the controlled resource as an argument
-		// For().
+		For(&certificatesv1.CertificateSigningRequest{}, builder.WithPredicates(certificateSigningRequestsPredicate)).
 		Complete(r)
 }
+
+// pendingCertFilter returns true if the CSR is pending approval
+func pendingCertFilter(obj runtime.Object) bool {
+	cert, ok := obj.(*certificatesv1.CertificateSigningRequest)
+	return ok && isPending(cert)
+}
+
+// reconcileCSR handles the CSR authorization and approval
+func (r *certificateSigningRequestsReconciler) reconcileCSR(request *certificatesv1.CertificateSigningRequest, instances *core.ConfigMap) error {
+	if request == nil || instances == nil {
+		return errors.New("CSR or instance ConfigMap cannot be nil")
+	}
+
+	// Get the list of instances that are expected to be Nodes
+	hosts, err := parseHosts(instances.Data)
+	if err != nil {
+		return errors.Wrapf(err, "unable to parse hosts from configmap")
+	}
+	var instanceAddresses []string
+	for _, host := range hosts {
+		instanceAddresses = append(instanceAddresses, host.Address)
+	}
+
+	certificateSigningRequest := csr.NewConfig(r.k8sclientset, instanceAddresses)
+
+	if authorize, err := certificateSigningRequest.Authorize(request); !authorize {
+		return errors.Wrapf(err, "could not authorize CSR %s", request.Name)
+	}
+	if err := certificateSigningRequest.Approve(request); err != nil {
```

Comment:
Whats the reasoning behind having Authorize and Approve as separate operations?
Is there a case where authorizing is required and approving is not? If not, it would make more sense for exported function in the CSR package to be Approve().